crates/lang/src/codegen/dispatch/execution.rs

```diff
@@ -70,24 +88,192 @@ pub struct ExecuteConstructorConfig {
+    ///
+    /// Since it is private it can only be used from within this crate.
+    /// We need this type in order to properly seal the `ConstructorReturnType`
+    /// trait from unwanted external trait implementations.
+    #[repr(transparent)]
+    pub struct Seal<T>(pub T);
+    impl<T> Sealed for Seal<T> {}
+}
+
+/// Guards against using invalid contract initializer types.
+///
+/// # Note
+///
+/// Currently the only allowed types are `()` and `Result<(), E>`
+/// where `E` is some unspecified error type.
+/// If the contract initializer returns `Result::Err` the utility
+/// method that is used to initialize an ink! smart contract will
+/// revert the state of the contract instantiation.
+pub trait ConstructorReturnType<C>: private::Sealed {
+    /// Is `true` if `Self` is `Result<C, E>`.
+    const IS_RESULT: bool = false;
+
+    /// The error type of the constructor return type.
+    ///
+    /// # Note
+    ///
+    /// For infallible constructors this is `core::convert::Infallible`.
+    type Error;
+
+    /// The type of the return value of the constructor.
+    ///
+    /// # Note
+    ///
+    /// For infallible constructors this is `()` whereas for fallible
+    /// constructors this is the actual return value. Since we only ever
+    /// return a value in case of `Result::Err` the `Result::Ok` value
+    /// does not matter.
+    type ReturnValue;
+
+    /// Converts the return value into a `Result` instance.
+    ///
+    /// # Note
+    ///
+    /// For infallible constructor returns this always yields `Ok`.
+    fn as_result(&self) -> Result<&C, &Self::Error>;
+
+    /// Returns the actual return value of the constructor.
+    ///
+    /// # Note
+    ///
+    /// For infallible constructor returns this always yields `()`
+    /// and is basically ignored since this does not get called
+    /// if the constructor did not fail.
+    fn return_value(&self) -> &Self::ReturnValue;
+}
+
+impl<C> ConstructorReturnType<C> for private::Seal<C> {
+    type Error = Infallible;
+    type ReturnValue = ();
+
+    #[inline(always)]
+    fn as_result(&self) -> Result<&C, &Self::Error> {
+        Ok(&self.0)
+    }
+
+    #[inline(always)]
+    fn return_value(&self) -> &Self::ReturnValue {
+        &()
+    }
+}
+
+impl<C, E> ConstructorReturnType<C> for private::Seal<Result<C, E>> {
+    const IS_RESULT: bool = true;
+    type Error = E;
+    type ReturnValue = Result<C, E>;
+
+    #[inline(always)]
+    fn as_result(&self) -> Result<&C, &Self::Error> {
+        self.0.as_ref()
+    }
+
+    #[inline(always)]
+    fn return_value(&self) -> &Self::ReturnValue {
+        &self.0
+    }
+}
+
+/// Trait used to convert return types of contract initializer routines.
+///
+/// Only `()` and `Result<(), E>` are allowed contract initializer return types.
+/// For `WrapReturnType<C>` where `C` is the contract type the trait converts
+/// `()` into `C` and `Result<(), E>` into `Result<C, E>`.
+pub trait InitializerReturnType<C>: private::Sealed {
```

Comment:
Since there are only two basic cases here, I don't know if I would go ahead with a new trait for this. I'd just do the wrapping in-line

Reply:
How would this look like?